src/gradient.c

```diff
@@ -80,3 +103,72 @@ void DrawHorizontalGradient(Drawable d, GC g,
 
    }
 }
+
+/** Draw a horizontal gradient. */
+void DrawHorizontalGradient(Drawable d, GC g,
+                            long fromColor, long toColor,
+                            int x, int y,
+                            unsigned int width, unsigned int height)
+{
+
+   const int shift = 15;
+   unsigned int column;
+   XColor colors[2];
+   int red, green, blue;
+   int ared, agreen, ablue;
+   int bred, bgreen, bblue;
+   int redStep, greenStep, blueStep;
+
+   /* Return if there's nothing to do. */
+   if(width == 0 || height == 0) {
+      return;
+   }
+
+   /* Here we assume that the background was filled elsewhere. */
+   if(fromColor == toColor) {
+      return;
+   }
+
+   /* Query the from/to colors. */
+   colors[0].pixel = fromColor;
+   colors[1].pixel = toColor;
+   JXQueryColors(display, rootColormap, colors, 2);
+
+   /* Set the "from" color. */
+   ared = (unsigned int)colors[0].red << shift;
+   agreen = (unsigned int)colors[0].green << shift;
+   ablue = (unsigned int)colors[0].blue << shift;
+
+   /* Set the "to" color. */
+   bred = (unsigned int)colors[1].red << shift;
+   bgreen = (unsigned int)colors[1].green << shift;
+   bblue = (unsigned int)colors[1].blue << shift;
+
+   /* Determine the step. */
+   redStep = (bred - ared) / (int)width;
+   greenStep = (bgreen - agreen) / (int)width;
+   blueStep = (bblue - ablue) / (int)width;
+
+   /* Loop over each line. */
+   red = ared;
+   blue = ablue;
+   green = agreen;
+
+   for(column = 0; column < width - 1; column++) {
+	  /* Determine the color for this line. */
+      colors[0].red = (unsigned short)(red >> shift);
+      colors[0].green = (unsigned short)(green >> shift);
+      colors[0].blue = (unsigned short)(blue >> shift);
+
+      GetColor(&colors[0]);
+
+      /* Draw the line. */
+      JXSetForeground(display, g, colors[0].pixel);
+      JXDrawLine(display, d, g, x + column, y, x + column + 1, y  + height - 1);
```

Comment:
Maybe I'm missing something, but I think the only differences between the horizontal and vertical gradients is the bounds of the for loop and the XDrawLine call. I think we could probably combine the two methods and save some duplication?

Reply:
Exactly in horizontal the code loops on columns (until width - 1 because the line must at last 1 point), in vertical it loops on lines.